// testcafe/lib/utils/browser-connection-timeouts.js
"use strict";
// -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHECK_STATUS_RETRY_DELAY = exports.SEND_STATUS_REQUEST_RETRY_COUNT = exports.SEND_STATUS_REQUEST_RETRY_DELAY = exports.SEND_STATUS_REQUEST_TIME_LIMIT = exports.CHECK_CHILD_WINDOW_DRIVER_LINK_DELAY = exports.CHECK_IFRAME_DRIVER_LINK_DELAY = exports.HEARTBEAT_INTERVAL = exports.BROWSER_RESTART_TIMEOUT = exports.HEARTBEAT_TIMEOUT = void 0;
exports.HEARTBEAT_TIMEOUT = 2 * 60 * 1000;
exports.BROWSER_RESTART_TIMEOUT = 60 * 1000;
exports.HEARTBEAT_INTERVAL = 2 * 1000;
exports.CHECK_IFRAME_DRIVER_LINK_DELAY = 500;
exports.CHECK_CHILD_WINDOW_DRIVER_LINK_DELAY = 500;
exports.SEND_STATUS_REQUEST_TIME_LIMIT = 5000;
exports.SEND_STATUS_REQUEST_RETRY_DELAY = 300;
exports.SEND_STATUS_REQUEST_RETRY_COUNT = Math.floor(exports.HEARTBEAT_TIMEOUT / exports.SEND_STATUS_REQUEST_RETRY_DELAY - 1);
exports.CHECK_STATUS_RETRY_DELAY = 1000;
