// testcafe/lib/browser/provider/built-in/locally-installed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const testcafe_browser_tools_1 = __importDefault(require("testcafe-browser-tools"));
exports.default = {
    openedBrowsers: {},
    isMultiBrowser: true,
    supportMultipleWindows: true,
    needCleanUpBrowserInfo: true,
    getActiveWindowId(browserId) {
        return this.openedBrowsers[browserId].activeWindowId;
    },
    setActiveWindowId(browserId, val) {
        this.openedBrowsers[browserId].activeWindowId = val;
    },
    cleanUpBrowserInfo(browserId) {
        delete this.openedBrowsers[browserId];
    },
    async openBrowser(browserId, pageUrl, browserName, allowMultipleWindows) {
        const args = browserName.split(' ');
        const alias = args.shift();
        const browserInfo = await testcafe_browser_tools_1.default.getBrowserInfo(alias);
        const openParameters = Object.assign({}, browserInfo);
        if (args.length)
            openParameters.cmd = args.join(' ') + (openParameters.cmd ? ' ' + openParameters.cmd : '');
        await testcafe_browser_tools_1.default.open(openParameters, pageUrl);
        let activeWindowId = null;
        if (allowMultipleWindows)
            activeWindowId = this.calculateWindowId();
        this.openedBrowsers[browserId] = { activeWindowId };
    },
    async isLocalBrowser() {
        return true;
    },
    async getBrowserList() {
        const installations = await testcafe_browser_tools_1.default.getInstallations();
        return Object.keys(installations);
    },
    async isValidBrowserName(browserName) {
        const browserNames = await this.getBrowserList();
        browserName = browserName.toLowerCase().split(' ')[0];
        return browserNames.indexOf(browserName) > -1;
    }
};
module.exports = exports.default;
