// testcafe-browser-tools/lib/exports.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var get_installations_1 = require("./api/get-installations");
exports.getInstallations = get_installations_1.default;
var find_window_1 = require("./api/find-window");
exports.findWindow = find_window_1.default;
var open_1 = require("./api/open");
exports.open = open_1.default;
var close_1 = require("./api/close");
exports.close = close_1.default;
var resize_1 = require("./api/resize");
exports.resize = resize_1.default;
var maximize_1 = require("./api/maximize");
exports.maximize = maximize_1.default;
var is_maximized_1 = require("./api/is-maximized");
exports.isMaximized = is_maximized_1.default;
var is_valid_device_name_1 = require("./api/is-valid-device-name");
exports.isValidDeviceName = is_valid_device_name_1.default;
var screenshot_1 = require("./api/screenshot");
exports.screenshot = screenshot_1.default;
var generate_thumbnail_1 = require("./api/generate-thumbnail");
exports.generateThumbnail = generate_thumbnail_1.default;
var get_browser_info_1 = require("./api/get-browser-info");
exports.getBrowserInfo = get_browser_info_1.default;
var get_viewport_size_1 = require("./api/get-viewport-size");
exports.getViewportSize = get_viewport_size_1.default;
var watch_window_1 = require("./api/watch-window");
exports.watchWindow = watch_window_1.default;
var bring_to_front_1 = require("./api/bring-to-front");
exports.bringWindowToFront = bring_to_front_1.default;
var get_devices_viewport_data_1 = require("./api/get-devices-viewport-data");
exports.getDevicesViewportData = get_devices_viewport_data_1.default;
const errors = __importStar(require("./errors"));
exports.errors = errors;
