// testcafe-legacy-api/lib/test-run-error/templates.js
'use strict';

exports.__esModule = true;

var _dedent = require('dedent');

var _dedent2 = _interopRequireDefault(_dedent);

var _lodash = require('lodash');

var _type = require('./type');

var _type2 = _interopRequireDefault(_type);

var _assertionDiffs = require('./assertion-diffs');

var _assertionDiffs2 = _interopRequireDefault(_assertionDiffs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function escapeNewLines(str) {
    return (0, _lodash.escape)(str).replace(/(\r\n|\n|\r)/gm, '\\n');
}

function getAssertionMsgPrefix(err) {
    return err.message ? `"${(0, _lodash.escape)(err.message)}" assertion` : 'Assertion';
}

function getDiffHeader(err) {
    if (err.isArrays) return `Arrays differ at index <code>${err.key}</code>:`;

    if (err.isObjects) return `Objects differ at the <code>${(0, _lodash.escape)(err.key)}</code> field:`;

    if (err.isStrings) return `Strings differ at index <code>${err.key}</code>:`;

    return '';
}

function markup(err, msgMarkup, opts = {}) {
    msgMarkup = (0, _dedent2.default)(`
        <div class="message">${(0, _dedent2.default)(msgMarkup)}</div>

        <strong>Browser:</strong> <span class="user-agent">${err.userAgent}</span>
    `);

    if (err.screenshotPath) msgMarkup += `\n<div class="screenshot-info"><strong>Screenshot:</strong> <a class="screenshot-path">${(0, _lodash.escape)(err.screenshotPath)}</a></div>`;

    if (!opts.withoutCallsite) {
        var callsiteMarkup = err.getCallsiteMarkup();

        if (callsiteMarkup) msgMarkup += `\n\n<strong>Code:</strong>\n${callsiteMarkup}`;
    }

    return msgMarkup;
}

exports.default = {
    [_type2.default.okAssertion]: err => markup(err, `
        ${getAssertionMsgPrefix(err)} failed at step "${(0, _lodash.escape)(err.stepName)}":

        <strong>Expected: </strong>not <code>null</code>, not <code>undefined</code>, not <code>false</code>, not <code>NaN</code> and not <code>''</code>
        <strong>Actual:   </strong><code>${escapeNewLines(err.actual)}</code>
    `),

    [_type2.default.notOkAssertion]: err => markup(err, `
        ${getAssertionMsgPrefix(err)} failed at step "${(0, _lodash.escape)(err.stepName)}":

        <strong>Expected: </strong><code>null</code>, <code>undefined</code>, <code>false</code>, <code>NaN</code> or <code>''</code>
        <strong>Actual:   </strong><code>${escapeNewLines(err.actual)}</code>
    `),

    [_type2.default.eqAssertion]: (err, maxStringLength) => {
        var diff = (0, _assertionDiffs2.default)(err, maxStringLength);
        var diffMarkerStr = diff.marker ? `          ${diff.marker}` : '';

        return markup(err, `
            ${getAssertionMsgPrefix(err)} failed at step "${(0, _lodash.escape)(err.stepName)}":

            ${getDiffHeader(err)}

            <strong>Expected: </strong><code>${escapeNewLines(diff.expected)}</code>
            <strong>Actual:   </strong><code>${escapeNewLines(diff.actual)}</code>
            <code>${diffMarkerStr}</code>
        `);
    },

    [_type2.default.notEqAssertion]: err => markup(err, `
        ${getAssertionMsgPrefix(err)} failed at step "${(0, _lodash.escape)(err.stepName)}":

        <strong>Expected: </strong>not <code>${escapeNewLines(err.actual)}</code>
        <strong>Actual:   </strong><code>${escapeNewLines(err.actual)}</code>
    `),

    [_type2.default.iframeLoadingTimeout]: err => markup(err, `
        IFrame loading timed out.
    `),

    [_type2.default.inIFrameTargetLoadingTimeout]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        IFrame target loading timed out.
    `),

    [_type2.default.uncaughtJSError]: err => {
        if (err.pageDestUrl) {
            return markup(err, `
                Uncaught JavaScript error <code>${(0, _lodash.escape)(err.scriptErr)}</code> on page <a href="${err.pageDestUrl}">${err.pageDestUrl}</a>
            `);
        }

        return markup(err, `
            Uncaught JavaScript error <code>${(0, _lodash.escape)(err.scriptErr)}</code> on page.
        `);
    },

    [_type2.default.uncaughtJSErrorInTestCodeStep]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        Uncaught JavaScript error in test code - <code>${(0, _lodash.escape)(err.scriptErr)}</code>.
    `),

    [_type2.default.storeDomNodeOrJqueryObject]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        It is not allowed to share the DOM element, jQuery object or a function between test steps via "this" object.
    `),

    [_type2.default.emptyFirstArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        A target element of the <code>${err.action}</code> action has not been found in the DOM tree.
        If this element should be created after animation or a time-consuming operation is finished, use the <code>waitFor</code> action (available for use in code) to pause test execution until this element appears.
    `),

    [_type2.default.invisibleActionElement]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        A target element <code>${(0, _lodash.escape)(err.element)}</code> of the <code>${err.action}</code> action is not visible.
        If this element should appear when you are hovering over another element, make sure that you properly recorded the <code>hover</code> action.
    `),

    [_type2.default.incorrectDraggingSecondArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>drag</code> action drop target is incorrect.
    `),

    [_type2.default.incorrectPressActionArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>press</code> action parameter contains incorrect key code.
    `),

    [_type2.default.emptyTypeActionArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The <code>type<code> action's parameter text is empty.
    `),

    [_type2.default.unexpectedDialog]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        Unexpected system <code>${err.dialog}</code> dialog <code>${(0, _lodash.escape)(err.message)}</code> appeared.
    `),

    [_type2.default.expectedDialogDoesntAppear]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The expected system <code>${err.dialog}</code> dialog did not appear.
    `),

    [_type2.default.incorrectSelectActionArguments]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>select</code> action's parameters contain an incorrect value.
    `),

    [_type2.default.incorrectWaitActionMillisecondsArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>wait</code> action's "milliseconds" parameter should be a positive number.
    `),

    [_type2.default.incorrectWaitForActionEventArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>waitFor</code> action's first parameter should be a function, a CSS selector or an array of CSS selectors.
    `),

    [_type2.default.incorrectWaitForActionTimeoutArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>waitFor</code> action's "timeout" parameter should be a positive number.
    `),

    [_type2.default.waitForActionTimeoutExceeded]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>waitFor</code> action's timeout exceeded.
    `),

    [_type2.default.incorrectGlobalWaitForActionEventArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>__waitFor</code> action's first parameter should be a function.
    `),

    [_type2.default.incorrectGlobalWaitForActionTimeoutArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>__waitFor</code> action's "timeout" parameter should be a positive number.
    `),

    [_type2.default.globalWaitForActionTimeoutExceeded]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>__waitFor</code> action's timeout exceeded.
    `),

    [_type2.default.emptyIFrameArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The selector within the <code>inIFrame</code> function returns an empty value.
    `),

    [_type2.default.iframeArgumentIsNotIFrame]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The selector within the <code>inIFrame</code> function doesn’t return an iframe element.
    `),

    [_type2.default.multipleIFrameArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The selector within the <code>inIFrame</code> function returns more than one iframe element.
    `),

    [_type2.default.incorrectIFrameArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        The <code>inIFrame</code> function contains an invalid argument.
    `),

    [_type2.default.uploadCanNotFindFileToUpload]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        Cannot find the following file(s) to upload:

        ${err.filePaths.map(path => `<code>${(0, _lodash.escape)(path)}</code>`).join(',\n')}
    `),

    [_type2.default.uploadElementIsNotFileInput]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>upload</code> action argument does not contain a file input element.
    `),

    [_type2.default.uploadInvalidFilePathArgument]: err => markup(err, `
        Error at step "${(0, _lodash.escape)(err.stepName)}":
        <code>upload</code> action's "path" parameter should be a string or an array of strings.
    `),

    [_type2.default.pageNotLoaded]: err => markup(err, `
        ${err.message}
    `)
};
module.exports = exports['default'];
